feat(curriculum): add case converter project to teach list comprehension (#52569)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-list-comprehension-by-building-a-case-converter-program/657effaa2a5e0277d71f9cbe.md

---
id: 657effaa2a5e0277d71f9cbe
title: Step 8
challengeType: 20
dashedName: step-8
---

# --description--

> By the end of the loop, `snake_cased_char_list` should contain all the converted characters in correct order. Use the `join()` string method to convert the list of characters into a string.

Reviewer: ```suggestion
By the end of the loop, `snake_cased_char_list` should contain all the converted characters in correct order. Use the `.join()` string method to convert the list of characters into a string.
```

```py
''.join(snake_cased_char_list)
```

> This joins the characters from the list to the empty string you called the `join()` method on. Save the resultant inside a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.

Reviewer 2: ```suggestion
This joins the characters from the list to the empty string you called the `join()` method on. Save the result of the `join()` call inside a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.
```

> This joins the characters from the list to the empty string you called the `join()` method on. Save the resultant inside a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.

Reviewer: ```suggestion
This joins the characters from the list to the empty string on which you called the `.join()` method. Save the result in a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.
```

---
id: 657effaa2a5e0277d71f9cbe
title: Step 8
challengeType: 20
dashedName: step-8
---

# --description--

By the end of the loop, `snake_cased_char_list` should contain all the converted characters in correct order. Use the `.join()` string method to convert the list of characters into a string.

```py
''.join(snake_cased_char_list)
```

This joins the characters from the list to the empty string on which you called the `.join()` method. Save the result in a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.

# --hints--

> You should join the characters in the list into a string using the `.join()` method and assign the string to a variable named `snake_cased_char_list`.

Reviewer 2: ```suggestion
You should join the characters in the list into a string using the `.join()` method and assign the string to a variable named `snake_cased_char_list` after your `for` loop.
```

```js
({
    test: () => {
        const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
        const convert_to_snake_case = __helpers.python.getDef("\n" + transformedCode, "convert_to_snake_case");
        const { function_body } = convert_to_snake_case;

>         assert.match(function_body, / +snake_cased_char_list\s*=\s*''\.join\(snake_cased_char_list\)/);

Reviewer 2: ```suggestion
        assert.match(function_body, / +snake_cased_char_list\s*=\s*["']{2}\.join\s*\(\s*snake_cased_char_list\s*\)/);
```

    }
})
```

# --seed--

## --seed-contents--

```py
def convert_to_snake_case(pascal_or_camel_cased_string):
    snake_cased_char_list = []
    for char in pascal_or_camel_cased_string:
        if char.isupper():

>             snake_cased_char_list.append(f'_{char.lower()}')

Reviewer 2: It would be better to restore the two lines version of this or change the step to use an f-string combined with the append call.

Author: Restored the two lines version

>     for char in pascal_or_camel_cased_string:
>         if char.isupper():
>             snake_cased_char_list.append(f'_{char.lower()}')
>         else:
>             snake_cased_char_list.append(char)
> --fcc-editable-region--
> 
> 
> --fcc-editable-region--

Reviewer 2: ```suggestion
--fcc-editable-region--
    for char in pascal_or_camel_cased_string:
        if char.isupper():
            snake_cased_char_list.append(f'_{char.lower()}')
        else:
            snake_cased_char_list.append(char)
--fcc-editable-region--
```

---
id: 657effaa2a5e0277d71f9cbe
title: Step 8
challengeType: 20
dashedName: step-8
---

# --description--

By the end of the loop, `snake_cased_char_list` should contain all the converted characters in correct order. Use the `.join()` string method to convert the list of characters into a string.

```py
''.join(snake_cased_char_list)
```

This joins the characters from the list to the empty string on which you called the `.join()` method. Save the result in a variable named `snake_cased_string` on the same level as the `snake_cased_char_list` variable.

# --hints--

> You should join the characters in the list into a string using the `.join()` method and assign the string to a variable named `snake_cased_char_list` after your `for` loop.

Reviewer 2: ```suggestion
You should join the characters in the `snake_cased_char_list` list into a string using the `.join()` method and assign the string to a variable named `snake_cased_string` after your `for` loop.
```

```js
({
    test: () => {
        const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
        const convert_to_snake_case = __helpers.python.getDef("\n" + transformedCode, "convert_to_snake_case");
        const { function_body } = convert_to_snake_case;

>         assert.match(function_body, / +snake_cased_char_list\s*=\s*["']{2}\.join\s*\(\s*snake_cased_char_list\s*\)/);

Reviewer 2: ```suggestion
        assert.match(function_body, / +snake_cased_string\s*=\s*["']{2}\.join\s*\(\s*snake_cased_char_list\s*\)/);
```
Wrong variable name.
